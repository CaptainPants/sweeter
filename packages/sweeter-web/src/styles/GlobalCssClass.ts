import { type StylesheetDependencyProvider } from './StylesheetDependencyProvider.js';
import { preprocessClassContent } from './preprocessor/preprocess.js';
import type {
    GlobalStyleSheetContentGeneratorContext,
    AbstractGlobalCssStylesheet,
    StylesheetGenerator,
    AbstractGlobalCssClass,
} from './types.js';

type GlobalCssClassContent =
    | ((self: GlobalCssClass) => string | StylesheetGenerator)
    | string;

type ContentConstructed = StylesheetGenerator | string;

export interface GlobalCssClassOptions {
    /**
     * Base the class name for this class on this value (it will be prefixed/suffixed or otherwise made unique).
     */
    className: string;
    content?: GlobalCssClassContent | undefined;
    preprocess?: boolean;
    extraDependencies?: StylesheetDependencyProvider;
}

export class GlobalCssClass
    implements AbstractGlobalCssStylesheet, AbstractGlobalCssClass
{
    public readonly className: string;
    public readonly id: string;
    public readonly symbol: symbol;
    public readonly classContent?: ContentConstructed | undefined;
    public readonly preprocess: boolean;

    #extraDependencies?: AbstractGlobalCssStylesheet[];

    constructor(options: GlobalCssClassOptions);
    constructor({
        className,
        content,
        preprocess = true,
        extraDependencies,
    }: GlobalCssClassOptions) {
        this.className = className;
        this.id = className;
        this.symbol = Symbol('GlobalCssClass-' + className);
        this.preprocess = preprocess;

        this.classContent =
            typeof content === 'function' ? content(this) : content;

        if (extraDependencies) {
            extraDependencies.addDependencyListener((dependency) => {
                this.#extraDependencies ??= [];
                this.#extraDependencies.push(dependency);
            });
        }
    }

    getContent(
        context: GlobalStyleSheetContentGeneratorContext,
    ): string | undefined {
        const name = context.getPrefixedClassName(this);

        const content =
            typeof this.classContent === 'string'
                ? this.classContent
                : this.classContent?.(context);

        if (!content) {
            return undefined;
        }

        const transformed = this.preprocess
            ? preprocessClassContent(name, content)
            : `.${name}\r\n{\r\n${content}\r\n}`;
        return transformed;
    }

    getReferencedStylesheets():
        | readonly AbstractGlobalCssStylesheet[]
        | undefined {
        let result: AbstractGlobalCssStylesheet[] | undefined =
            typeof this.classContent === 'function'
                ? [...this.classContent.references]
                : undefined;

        if (this.#extraDependencies) {
            result ??= [];
            result.push(...this.#extraDependencies);
        }
        return result;
    }
}
