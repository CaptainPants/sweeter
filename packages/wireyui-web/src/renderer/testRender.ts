import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
