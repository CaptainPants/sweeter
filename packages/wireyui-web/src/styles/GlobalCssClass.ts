import type { GlobalStyleSheet } from './types.js';

let counter = 0;

export class GlobalCssClass implements GlobalStyleSheet {
    public readonly className: string;
    public readonly content: string;
    public readonly id: symbol;

    constructor(options: { id: string; content: string }) {
        this.className = '_glbl' + counter + '_' + options.id;
        ++counter;

        this.content = `.${this.className} {\r\n${options.content}\r\n}`;
        this.id = Symbol('GlobalCssClass');
    }
}
