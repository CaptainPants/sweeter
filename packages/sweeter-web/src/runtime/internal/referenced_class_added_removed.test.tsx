/* @jsxImportSource ../.. */

import { testRender } from '../../test/testRender.js';
import { GlobalCssClass } from '../../styles/GlobalCssClass.js';

const class_ = new GlobalCssClass({
    className: 'Test1',
    content: `
        color: green;

        &:hover {
            color: red;
        }
    `,
});

it('added and removed', () => {
    const res = testRender(() => <div class={class_}>TEXT</div>);

    // class 'class_' should have been added to the document.head
    expect(document.head.outerHTML).toMatchSnapshot();

    res.dispose();

    // now it should have been removed
    expect(document.head.outerHTML).toMatchSnapshot();
});
