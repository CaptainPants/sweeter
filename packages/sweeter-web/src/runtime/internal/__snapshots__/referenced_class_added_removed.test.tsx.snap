// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`added and removed 1`] = `
"<head><style data-id=\\"Test1\\">
._glbl0_Test1 {
    color: green;
}
._glbl0_Test1:hover {
    color: red;
}
</style></head>"
`;

exports[`added and removed 2`] = `"<head></head>"`;
